packages/beacon-node/src/chain/archiveStore/utils/hierarchicalLayers.ts

```diff
@@ -0,0 +1,128 @@
+
+export type Layers = {
+  snapshotSlot: Slot;
+  diffSlots: Slot[];
+};
+
+export class HierarchicalLayers {
+  private snapshotEverySlot: number;
+  private diffEverySlot: number[];
+
+  /**
+   * Initialized with the comma separated values in ascending order e.g. 2,4,6,10
+   * These values will represent every nth epoch and each consider as a layer
+   * The last value which should be highest should be consider as snapshot layer.
+   */
+  constructor(epochs: Epoch[]) {
+    const lastEpoch = epochs.at(-1);
+    if (!lastEpoch) throw new Error("Must provide a list of epochs");
+    this.snapshotEverySlot = lastEpoch * SLOTS_PER_EPOCH;
+
+    this.diffEverySlot = epochs
+      .slice(0, -1)
+      // Reverse here, so lower layer get higher priority when matching
+      .reverse()
+      .map((s) => s * SLOTS_PER_EPOCH);
+  }
+
+  static fromString(layers: string = DEFAULT_DIFF_LAYERS) {
+    const layerEpochs = [
+      ...new Set(
+        layers
+          .split(",")
+          .map((s) => s.trim())
+          .map((n) => parseInt(n))
+      ),
+    ];
+
+    if (layerEpochs.length !== layers.split(",").length) {
+      throw new Error(`Please provide unique epoch intervals. Given = ${layers}`);
+    }
+
+    if ([...layerEpochs].sort((a, b) => a - b).join(",") !== layerEpochs.join(",")) {
+      throw new Error(`Please provide diff layers in ascending order. Given = ${layers}`);
+    }
+
+    return new HierarchicalLayers(layerEpochs);
+  }
+
+  toString(): string {
+    return `${this.diffEverySlot
+      .reverse()
+      .map((s) => s / SLOTS_PER_EPOCH)
+      .join(",")},${this.snapshotEverySlot / SLOTS_PER_EPOCH}`;
+  }
+
+  get totalLayers(): number {
+    return this.diffEverySlot.length + 1;
+  }
+
+  getStorageType(slot: Slot, archiveMode: ArchiveMode): HistoricalStateStorageType {
+    if (archiveMode === ArchiveMode.Frequency) return HistoricalStateStorageType.Full;
+
+    if (slot === 0) {
+      return HistoricalStateStorageType.Snapshot;
+    }
+
+    if (slot % this.snapshotEverySlot === 0) return HistoricalStateStorageType.Snapshot;
+    if (this.diffEverySlot.some((s) => slot % s === 0)) return HistoricalStateStorageType.Diff;
+
+    return HistoricalStateStorageType.BlockReplay;
+  }
+
+  getArchiveLayers(slot: Slot): Layers {
+    const path: Slot[] = [];
+    let lastSlot: number | undefined = undefined;
+
+    for (let layer = 0; layer < this.totalLayers; layer++) {
+      const newSlot = this.getPreviousSlotForLayer(slot, layer);
+      if (lastSlot === undefined || newSlot > lastSlot) {
+        lastSlot = newSlot;
+        path.push(newSlot);
+      }
+    }
+
+    const diffSlots = [...new Set(path)];
+    const snapshotSlot = diffSlots.shift();
+
+    if (snapshotSlot == null) {
+      throw new Error(`Can not find snapshot layer for slot=${slot}`);
+    }
+
+    return {
+      snapshotSlot,
+      diffSlots,
+    };
+  }
+
+  getPreviousSlotForLayer(slot: Slot, layer: number): Slot {
```

Comment:
Took me a while to understand what is expected from `layer`. 

So when `layer == 0` it really refers to the snapshot layer, `layer == 1` refers to lowest frequent differential layer (ie. `this.diffEverySlot[layer - 1`). 

Need to add comment to clarify what `layer` really means here and what value represents what